File: modules/player/data/Player.ts
import {BaseData, StateData} from "../../core/data/BaseData";
import {DataOccasion, dataPK, field, occasion} from "../../core/data/DataLoader";
import {DynamicData} from "../../core/data/DynamicData";
import {LevelCalculator} from "../utils/LevelCalculator";

const DefaultGold = 100;

export type WxUserInfo = {
	nickName: string,
	avatarUrl: string
}

export enum PlayerState {
	Newer, Normal, Banned
}

export interface IPlayerData {
	openid: string;
}

export type PlayerEditableInfo = {
	name?: string
	gender?: number
	nickName?: string
	phone?: string
	avatarUrl?: string
	slogan?: string
}

export class PlayerBaseInfo extends BaseData {

	@field(String)
	public openid: string
	@field(Number)
	public chainId: number
	@field(String)
	public address: string
	@field(String)
	public name: string
	@field(String)
	public nickName: string
	@field(String)
	public avatarUrl: string
	@field(Number)
	public level: number
	@field
	public slogan: string
}

export class Player extends StateData<PlayerState> {

	@field @dataPK
	public openid: string;
	@field(Number)
	public chainId: number;
	@field
	public address: string; // 钱包地址

	@field
	public name: string = "";
	@field(Number)
	public gender: number = 0;
	@field
	public nickName: string = ""; // 微信名称
	@field(String)
	public phone: string;
	@field
	public avatarUrl: string = "";
	@field
	public slogan: string = "";
	@field
	public description: string = "";
	// @field
	// public level: number = 1;
	@field
	public exp: number = 0;

	@field
	public score: number = 0; // 贡献分

	@field(String)
	public inviteCode: string;

	@field(Number)
	public createTime: number;

	@field(Number)
	public gold: number = DefaultGold;

	@field(Number)
	public state: PlayerState = PlayerState.Newer;

	// region 额外数据

	public refreshExtra() {

	}

	// endregion

	// region 金币控制

	/**
	 * 获得金钱
	 */
	public gainGold(val) {
		this.gold = Math.max(0, this.gold + val);
	}

	// endregion

	// region 等级控制

	@field(Number)
	@occasion(DataOccasion.Extra)
	public get level() {
		return LevelCalculator.level(this.exp);
	}
	@field(Number)
	@occasion(DataOccasion.Extra)
	public get curExp() {
		return LevelCalculator.curExp(this.exp);
	}
	@field(Number)
	@occasion(DataOccasion.Extra)
	public get restExp() {
		return LevelCalculator.restExp(this.exp);
	}
	@field(Number)
	@occasion(DataOccasion.Extra)
	public get expRate() {
		return LevelCalculator.expRate(this.exp);
	}

	/**
	 * 获得经验
	 */
	public gainExp(val) {
		this.exp = Math.max(0, this.exp + val);
		this.refresh();
	}

	/**
	 * 刷新等级
	 */
	protected _level;
	private refreshLevel() {
		const level = this.level;
		if (level > this._level) // 下一级
			this.onLevelUp(this._level, level);
		this._level = level
	}

	/**
	 * 升级回调
	 */
	protected onLevelUp(oldLevel, newLevel) { }

	// endregion

	// region 积分控制

	public gainScore(score) {
		this.score += score;
	}

	// endregion

	// region 刷新

	public refresh() {
		this.refreshLevel();
		this.refreshExtra();
	}

	// endregion

	// region 业务操作

	/**
	 * 修改数据
	 */
	public edit(info: PlayerEditableInfo) {
		Object.assign(this, info);
	}

	// endregion
}

export abstract class PlayerData extends DynamicData implements IPlayerData {

	@field(String)
	@dataPK
	public openid: string;

}

File: Learning/project/Go/homi-server/model/playerdata/player.go
package playerdata

import (
	"math"
)

// 状态常量
const (
	NewerPlayer  int8 = 0
	NormalPlayer int8 = 1
	BannedPlayer int8 = 2
)

const (
	DefaultGold = 200
	DefaultExp  = 0
)

type PlayerBaseInfo struct {
	PlayerData `json:",inline" bson:",inline"`
	ChainId    int64  `json:"-" bson:"chainId"`
	Address    string `json:"-" bson:"address"` //钱包地址
	Name       string `json:"name" bson:"name"`
	Gender     int8   `json:"gender" bson:"gender"`
	NickName   string `json:"nickName" bson:"nickName"`
	Phone      string `json:"-" bson:"phone"`
	AvatarUrl  string `json:"avatar" bson:"avatar"`
	//Level     int    `json:"level" bson:"level"`
	Slogan string `json:"slogan" bson:"slogan"`
}

type Player struct {
	PlayerBaseInfo `json:",inline" bson:",inline"`
	Description    string `json:"description" bson:"description"`
	Exp            int    `json:"exp" bson:"exp"`
	Score          int    `json:"score" bson:"score"`
	InviteCode     string `json:"inviteCode" bson:"inviteCode"`
	Gold           int    `json:"gold" bson:"gold"`

	CreateTime int64 `json:"createTime" bson:"createTime"`

	//StateData
	State     int8   `json:"state" bson:"state"`
	StateTime int64  `json:"stateTime" bson:"stateTime"`
	StateDesc string `json:"stateDesc" bson:"stateDesc"`
}

type PlayerEditableInfo struct {
	Name      string `json:"name" bson:"name"`
	Gender    int8   `json:"gender" bson:"gender"`
	NickName  string `json:"nickName" bson:"nickName"`
	Phone     string `json:"-" bson:"phone"`
	AvatarUrl string `json:"avatar" bson:"avatar"`
	Slogan    string `json:"slogan" bson:"slogan"`
}

//func (u Player) GetID() string {
//	return u.ID
//}
//
//func (u Player) GetCollection() *mongo.Collection {
//	return oprs.MongoOpr.Collection("Player")
//}
//
//func (u Player) Save() error {
//	return myMongo.Save(u)
//}

func (u *Player) GainExp(exp float64) {
	u.Exp = int(math.Ceil(exp)) + u.Exp
}

func (u *Player) GainGold(gold float64) {
	//filter := bson.D{{"_id", u.ID}}
	//gainGold := bson.D{{"$set", bson.D{{"gold", int(math.Ceil(gold)) + u.Gold}}}}
	//_, err := u.GetCollection().UpdateOne(context.TODO(), filter, gainGold)
	u.Gold = int(math.Ceil(gold)) + u.Gold
}

func (u *Player) ConsumeGold(price int) {
	u.Gold -= price
	//err := u.Save()
	//if err != nil {
	//	return err
	//}
	//return nil
}

//func GetBaseInfo(openid string) (*PlayerBaseInfo, error) {
//	player :=
//	err := myMongo.FindOne(player.GetCollection(), "openid", openid, baseResult)
//	if err != nil {
//		return nil, err
//	}
//	return baseResult, nil
//}

//region mysql相关操作

//func (u *Player) BeforeCreate(tx *gorm.DB) (err error) {
//	node, err := snowflake.NewNode(1) //
//	u.ID_ = int64(node.Generate())
//	u.State = NewerPlayer
//	return err
//}
//
//func (u *Player) TableName() string {
//	return "player"
//}

//endregion
